Introduce NVIDIA package

rene opened the pull request:
## Description

This PR introduces the pkg/nvidia package. This package fetches the Jetson Linux (currently 35.5.0) and extracts all .deb  packages containing all relevant files (libraries, firmwares, etc) needed to run specific NVIDIA GPU applications, such as CUDA applications. Then, all provided CDI files are parsed by the process-cdi.sh script, which will copy all files pointed in the CDI from the extracted .deb packages.
    
The output container will provide a directory /opt/nvidia, where:

- /opt/vendor/nvidia/dist: Contains all files pointed in all CDI files
- /opt/vendor/nvidia/bin/nvidia-ctk and
- /opt/vendor/nvidia/bin/nvidia-cdi-hook: applications needed to process libraries files from the CDI spec.
- /opt/vendor/nvidia/bin/ldconfig-glibc: ldconfig tool for GNU libc
- /opt/vendor/nvidia/init.d/nv-init.sh: Script to perform platform setup actions, to be executed during pillar's initialization

All supported CDI files will be available at /etc/cdi.
    
The following CDI files are provided:

- jetson-xavier-nx.yaml: For devices based on Jetson Xavier NX
- jetson-orin-nano.yaml: For devices based on Jetson Orin Nano

## Notes

- The hash for pkg/alpine will be updated across other packages in a separated PR, once this gets approved.
- I've opened a ticket on LF's Jira for the creation of eve-nvidia package under LF's dockerhub

rene commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,102 @@
+
+# We need to build dpkg in order to get zstd compression support
+ENV DPKG_REVISION=1.22.6
+ADD --keep-git-dir=true https://salsa.debian.org/dpkg-team/dpkg.git#${DPKG_REVISION} /dpkg
```
FWIW, AFAIK there is no way to get rid of this error, since it's not a rule, it's hadolint that doesn't recognize `--keep-git-dir` as a valid flag...

christoph-zededa replied:
You could do without `--keep-git-dir` and then before `git am`: `git init; git add .; git commit -m "initial commit"`

But IMO it is also fine to just add another hadolint-ignore line.

rene replied:
I prefer to keep the `--keep-git-dir` line as we use in many other Dockerfile across pkg/. However, as I said, it's not a hadolint's rule, it doesn't recognize the flags, so unfortunately there is no `hadolint-ignore` to use...

christoph-zededa replied:
Fine.

rene commented:
Regarding the SPDX for CDI files: I can't claim the license of these files, they were auto-generated (with minimal manual changes from myside) by nvidia-container-toolkit.

christoph-zededa commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,102 @@
+        git am "$x" ; \
+    done
+
+RUN mkdir -p dist && \
+    make cmd-nvidia-ctk && cp nvidia-ctk dist/ && \
+    make cmd-nvidia-cdi-hook && cp nvidia-cdi-hook dist/
```
```suggestion
RUN mkdir -p dist
RUN make cmd-nvidia-ctk && cp nvidia-ctk dist/
RUN make cmd-nvidia-cdi-hook && cp nvidia-cdi-hook dist/
```
not everything has to be in one `RUN` - this way it is more readable (subjectively) and it can cache more (objectively)

rene replied:
yeah, I agree, in this case it's better to break into multiple RUNs... it's done.

OhmSpectator commented:
> Regarding the SPDX for CDI files: I can't claim the license of these files, they were auto-generated (with minimal manual changes from myside) by nvidia-container-toolkit. We are good to go, files were auto-generated and there isn't any license restriction explicitly informed + the tool is intended to be used to generate these files along Linux distros.

Maybe we can just mark the entire /pkg/nvidia/cdi directory as ignored by the SPDX check? Like we do with all vendors dir...

rene commented:
> > Regarding the SPDX for CDI files: I can't claim the license of these files, they were auto-generated (with minimal manual changes from myside) by nvidia-container-toolkit. We are good to go, files were auto-generated and there isn't any license restriction explicitly informed + the tool is intended to be used to generate these files along Linux distros.
> 
> Maybe we can just mark the entire /pkg/nvidia/cdi directory as ignored by the SPDX check? Like we do with all vendors dir...

@OhmSpectator , makes sense, even better: we could create an ignore file like .gitignore, and add everything that must be ignored inside, what do you think?

OhmSpectator commented:
> > > Regarding the SPDX for CDI files: I can't claim the license of these files, they were auto-generated (with minimal manual changes from myside) by nvidia-container-toolkit. We are good to go, files were auto-generated and there isn't any license restriction explicitly informed + the tool is intended to be used to generate these files along Linux distros.
> > 
> > 
> > Maybe we can just mark the entire /pkg/nvidia/cdi directory as ignored by the SPDX check? Like we do with all vendors dir...
> 
> @OhmSpectator , makes sense, even better: we could create an ignore file like .gitignore, and add everything that must be ignored inside, what do you think?

Good idea. I can implement the logic and we can test it on your PR ))

OhmSpectator commented:
@rene, here is the PR that should help: https://github.com/lf-edge/eve/pull/4190

shjala commented:
> files were auto-generated and there

Would you please include the scripts used for the cid yaml files in /tools? makes it easier for others to make changes.

shjala commented on pkg/nvidia/build.yml:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+org: lfedge
+image: eve-nvidia
```
Can we get this to not run as root? if you think it is possible I can tell what changes are needed.

OhmSpectator replied:
BTW, shouldn't we also create a repo on dockerhub where the image will be published?
https://hub.docker.com/r/lfedge/eve-nvidia does not exist...

rene replied:
@shjala , this image doesn't run, it's just unpacked to the rootfs as it's included to the init section of the image yaml file (through the nvidia modifier). The only executable script (nv-init.sh) it's executed by pillar during initialization, so it runs on pillar's context...

@OhmSpectator , I've opened a ticket already at LF's Jira in order to create this repository under docker hub, so it should be ready soon.

OhmSpectator replied:
ah, sorry, didn't notice it

shjala replied:
ah :(

shjala replied:
so how are we going to run apps? in context of pillar process?

rene replied:
That's how Pillar's initiate custom vendor stuff.... it runs anything executable under `/opt/vendor/*/init.d` during initialization: https://github.com/lf-edge/eve/blob/master/pkg/pillar/scripts/device-steps.sh#L305

rene commented:
> > files were auto-generated and there
> 
> Would you please include the scripts used for the cid yaml files in /tools? makes it easier for others to make changes.

Which one do you mean, process-cdi.sh? This is very NVIDIA specific, I think it's better to keep it inside the pkg/nvidia....

shjala commented:
> > > files were auto-generated and there
> > 
> > 
> > Would you please include the scripts used for the cid yaml files in /tools? makes it easier for others to make changes.
> 
> Which one do you mean, process-cdi.sh? This is very NVIDIA specific, I think it's better to keep it inside the pkg/nvidia....

I mean the script that generated the `eve/pkg/nvidia/cdi/*.yml` files.

rene commented:
> > > > files were auto-generated and there
> > > 
> > > 
> > > Would you please include the scripts used for the cid yaml files in /tools? makes it easier for others to make changes.
> > 
> > 
> > Which one do you mean, process-cdi.sh? This is very NVIDIA specific, I think it's better to keep it inside the pkg/nvidia....
> 
> I mean the script that generated the `eve/pkg/nvidia/cdi/*.yml` files.

There are no scripts for that, these files are generated in a bare-metal Jetpack Linux using the nvidia-ctk tool (from the nvidia-container-toolkit), here these files are just post-processed in order to fix all the paths of the Jetpack libraries, which on EVE are placed under the /opt/vendor/nvidia/.... directory. I will update the commit message to contain this information...

christoph-zededa commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,102 @@
+# Get and extract ldconfig for glibc
+ADD http://ports.ubuntu.com/ubuntu-ports/pool/main/g/glibc/libc-bin_2.31-0ubuntu9.16_arm64.deb /
+RUN mkdir -p /ldconfig-bin && \
+    dpkg -x /libc-bin_2.31-0ubuntu9.16_arm64.deb /ldconfig && \
+    cp /ldconfig/sbin/ldconfig.real /ldconfig-bin/ldconfig-glibc
```
```suggestion
RUN ar x libc-bin_2.31-0ubuntu9.16_arm64.deb.tar.xz
RUN tar xvf data.tar.xz ./sbin/ldconfig
RUN mv ./sbin/ldconfig.real /ldconfig-bin/ldconfig-glibc
```

no `dpkg` necessary

rene replied:
Yeah, but I still prefer to keep using dpkg because as soon as we migrate to Alpine 3.20 it won't be needed to build dpkg anymore....

OhmSpectator commented:
@rene, since #4190 is merged, please rebase on master and add the CDI files to SPDX ignore file.

rene commented:
Updates in this PR:
- Rebased onto master
- Add CDI directory to .spdxignore file
- Add information about how CDI files were generated to the commit message

OhmSpectator commented:
> * Add CDI directory to .spdxignore file

Yay! Thanks! =)

eriknordmark reviewed (commented):
What's the impact of /etc/cdi and /opt/vendor on the size of the EVE image?
And do you build and include this for x86 as well? I didn't see where it gets excluded.

rene commented:
> What's the impact of /etc/cdi and /opt/vendor on the size of the EVE image? And do you build and include this for x86 as well? I didn't see where it gets excluded.

For the /etc/cdi it's roughly 244KB, the rootfs image it's around 311MB, so almost 100MB bigger than a regular image. However, these files are only shipped with nvidia platform variant, they are not included on generic/imx8mp images (this is done using the yaml modifier file, that only includes the pkg/nvidia image for NVIDIA platform). Also, only the arm64 pkg/nvidia image contains all these files, they are not include on riscv64/amd64 architectures, see https://github.com/lf-edge/eve/pull/4189/files#diff-cdeb471c1ec66a79874c91a3ee84f49bff00982b28c03c729ded9c510eb85e7cR11 .

shjala commented:
> Approve; there is one Yetus complain.

That's an issue in Hadolint, see https://github.com/hadolint/hadolint/issues/951

rene commented:
> > Approve; there is one Yetus complain.
> 
> That's an issue in Hadolint, see [hadolint/hadolint#951](https://github.com/hadolint/hadolint/issues/951)

Yeah, thanks @shjala for the link

deitch commented:
> Add information about how CDI files were generated to the commit message

Shouldn't that be in `pkg/nvidia/README.md` as well? Person is most likely to look at it in the code, rather than looking for a commit.

deitch commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,102 @@
+
+FROM scratch
+COPY --from=build /cdi-dist/*.yaml /etc/cdi/
+COPY --from=build /rules.d/* /opt/vendor/nvidia/etc/udev/rules.d/
```
In most other packages, we have the build image install lots of things to `/out/` and then have a single line in 

```dockerfile
FROM scratch
COPY --from=build /out /
```

Would that work here?

rene replied:
Probably, but I'm just following the same approach of many other packages, like kvm-tools, bsp-imx, newlog, bpftrace, gpt-tools, udev, wwan, etc... so I would let as it is...

deitch commented on pkg/nvidia/scripts/process-cdi.sh:
```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
```
So is the purpose of `process-cdi.sh` to be run during `pkg/nvidia` build (i.e. `lfedge/eve-nvidia` image), when it reads the CDI yaml files, and uses that to find any binaries it needs in the `.deb` files and place them in the final image, so they are available to pillar when needed?

rene replied:
Exactly this.

deitch replied:
Can we mark that, both here and in the README? So people can glance at it and know how things work?

deitch commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,102 @@
+
+ADD ${JETSON_LINUX} /jetson_linux.tbz2
+
+# We need to build dpkg in order to get zstd compression support
```
Two questions:

1. I didn't understand this. Why do we need dpkg to get zstd compression support?
2. Why don't we just use the [alpine dpkg package](https://pkgs.alpinelinux.org/package/edge/main/aarch64/dpkg) (that sounds funny when you say it out loud)?

Or are you saying not that we need dpkg in order to use zstd on other things, but if a `.deb` package is compressed with zstd, then we need _dpkg itself_ to support zstd, and the version packaged with alpine does not? And so we need to compile it?

deitch replied:
Ah, and Alpine 3.20 has support for it, hence your comment below?

deitch replied:
Can we put that in the comment here, so it is clear? That the `.deb` we will unpack might have zstd compression support, and our Alpine package dpkg is from alpine 3.16 and does not support it, and when we switch to alpine 3.20 we can get rid of all of this? And maybe a TODO line as well?

deitch commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,102 @@
+RUN make cmd-nvidia-ctk && cp nvidia-ctk dist/
+RUN make cmd-nvidia-cdi-hook && cp nvidia-cdi-hook dist/
+
+# Get and extract ldconfig for glibc
```
Why do we need ldconfig for glibc?

rene replied:
This one runs over the rootfs of the application container, like CUDA application containers (glibc based), not over the host....

rene commented:
> > Add information about how CDI files were generated to the commit message
> 
> Shouldn't that be in `pkg/nvidia/README.md` as well? Person is most likely to look at it in the code, rather than looking for a commit.

I agree, I will add a proper README.md file.

rene commented:
Updates in this PR:
- Rebased onto master
- Addressed reviewers comments:
  - Added a README.md to pkg/nvidia
  - Added relevant comments to Dockerfile and the process-cdi.sh script

deitch commented on .spdxignore:
```diff
@@ -2,8 +2,9 @@ pkg/edgeview/vendor/
 pkg/kube/eve-bridge/vendor/
```
Not relevant to this PR, but can we have a more general pattern for ignoring vendor?

rene replied:
@OhmSpectator , any comments?

OhmSpectator replied:
Yeah...
In theory, we can put just /vendor/ there... In this case, a file will be ignored if it has "/vendor/" as a part of its path.

deitch commented on images/modifiers/nvidia.yq:
```diff
@@ -0,0 +1 @@
+.init += ["NVIDIA_TAG"]
```
Nice. This is a whole lot easier to understand with the `image/modifiers/` structure.

deitch reviewed (approved):
I like this. Easy to find the README and start with it, and then everything fits. Thanks @rene .

eriknordmark commented:
> There are no scripts for that, these files are generated in a bare-metal Jetpack Linux using the nvidia-ctk tool (from the nvidia-container-toolkit), here these files are just post-processed in order to fix all the paths of the Jetpack libraries, which on EVE are placed under the /opt/vendor/nvidia/.... directory. I will update the commit message to contain this information...

In addition to the commit message it would make sense to have markdown file which documents how to update to new jetpack versions or add support for new GPUs by describing the necessary steps. Is there already a markdown file which talks about NVIDIA GPU handling?

FWIW I see that lfedge/eve-nvidia is created so otherwise this looks good to go (to me).

rene commented:
> > There are no scripts for that, these files are generated in a bare-metal Jetpack Linux using the nvidia-ctk tool (from the nvidia-container-toolkit), here these files are just post-processed in order to fix all the paths of the Jetpack libraries, which on EVE are placed under the /opt/vendor/nvidia/.... directory. I will update the commit message to contain this information...
> 
> In addition to the commit message it would make sense to have markdown file which documents how to update to new jetpack versions or add support for new GPUs by describing the necessary steps. Is there already a markdown file which talks about NVIDIA GPU handling?
> 
> FWIW I see that lfedge/eve-nvidia is created so otherwise this looks good to go (to me).

@eriknordmark , I've added more information to the README.md file as you requested. Also, as you pointed out, the eve-nvidia package is ready on dockerhub, so we are good to go.

eriknordmark reviewed (approved):
LGTM
And ignoring the yetus complaint about 
hadolint: invalid flag: --keep-git-dir

One last question: what is the impact on the image size before and after this PR?

rene commented:
> LGTM And ignoring the yetus complaint about hadolint: invalid flag: --keep-git-dir
> 
> One last question: what is the impact on the image size before and after this PR?

@eriknordmark , for non-nvidia images the impact on the image size is **none**, because no pkg/nvidia will be included. For nvidia platform images, it will increase in about 100MB due to all Jetson libraries, files, etc, that will be part of the rootfs image...

OhmSpectator commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,105 @@
+
+RUN eve-alpine-deploy.sh
+
+# This is an arm64 (only) package, nothing to do for other architectures
```
Not critical, but is there a way to avoid running most of this file if we are not doing ARM build? Just the build packages installation takes ~10 seconds of time....

rene replied:
Yeah, I can move the RUN eve-alpine-deploy.sh to the arm64 container....

OhmSpectator replied:
If it's a one-minute change, it would be nice to have it. If it's complicated and will require careful retesting, I can survive 10 extra seconds spent during the build.

rene replied:
done.

OhmSpectator commented on pkg/nvidia/README.md:
```diff
@@ -0,0 +1,90 @@
+should be considered:
+
+1. Ensure EVE's NVIDIA custom kernel is compatible with the new Jetpack version
+1. Update NVIDIA firmwares on pkg/fw (if required)
```
I like that it renders to a properly numbered list, nevertheless.

![image](https://github.com/user-attachments/assets/fe13e2f5-9ac7-40df-9aad-4c56e761ba52)

I did not know about this feature)

OhmSpectator commented on pkg/nvidia/scripts/process-cdi.sh:
```diff
@@ -0,0 +1,56 @@
+    echo "Use: $0 <cdi-file> <rootfs-extracted> <destination-folder>"
+    echo
+    exit 1
+else
```
I don't think we need `else` here, as the previous branch just ends with exit =)

OhmSpectator commented on Makefile:
```diff
@@ -378,8 +378,13 @@ ifeq ($(HV),kubevirt)
 else
         #kube container will not be in non-kubevirt builds
         PKGS_$(ZARCH)=$(shell find pkg -maxdepth 1 -type d | grep -Ev "eve|alpine|sources|kube|external-boot-image|verification$$")
+# nvidia platform requires more space
```
It would be nice to indent it for readability.

OhmSpectator replied:
I would suggest keeping the same style as in the rest of the file and maybe addressing the entire file later, but definitely not in this PR...

rene replied:
makes sense. it's done.

OhmSpectator reviewed (approved):
Looks nice.
I have only minor comments, that can be addressed only if you have a will.

OhmSpectator reviewed (approved):
Perfect

christoph-zededa commented on pkg/nvidia/Dockerfile:
```diff
@@ -0,0 +1,105 @@
+# Copyright (c) 2024 Zededa, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+FROM lfedge/eve-alpine:82df60e43ab9f8c935584b8c7b4d0a4b0271d608 as build-base
```
What about bumping the other packages as well?
```
find ./ -name Dockerfile -exec perl -pnibak -e 's/lfedge\/eve-alpine(\S+)/lfedge\/eve-alpine:82df60e43ab9f8c935584b8c7b4d0a4b0271d608/' "{}" \;
```

rene replied:
As stated in the PR's description, it's going to be done once this one is approved. Now that is approved, I will open another PR to bump up all other packages...

christoph-zededa replied:
Thx, fine.

christoph-zededa commented on pkg/nvidia/cdi/jetson-xavier-nx.yaml:
```diff
@@ -0,0 +1,3355 @@
+---
```
Do you want perhaps add a comment on how you created these files?

christoph-zededa replied:
IMO would be better to have comments near the code, not far away. A link would be ok, too.

> that's why the post-processing is required to adjust pathnames of libraries, firmwares and other binary files

How is this done?

rene replied:
I want to change the auto-generated yaml file as minimal as possible. The whole explanation is on the README.md file from the package, so I prefer to keep as it is....
